New provider provisioning for RHV

borod108 opened the pull request:
Use OvirtSDK for the provisioning flow, this respects use_ovirt_engine_sdk setting
and will not use OvirtSDK if it is off.

~This PR is based on: https://github.com/ManageIQ/manageiq/pull/14399~

pkliczewski commented on app/models/manageiq/providers/redhat/infra_manager/refresh/strategies/api3.rb:
```diff
@@ -1,4 +1,37 @@
   class Api3 < ManageIQ::Providers::Redhat::InfraManager::Refresh::Refresher
+    def host_targeted_refresh(inventory, target)
+      methods = {
+        :primary => {
```
We have a bug with host targeted refresh and we are missing:

          :network => {:networks => "network" },

durandom commented:
@borod108 after rebasing, could you also run `rubocop -a` on the reported files?

durandom commented on app/models/manageiq/providers/redhat/infra_manager/ovirt_services/strategies/v3.rb:
```diff
@@ -19,5 +25,185 @@ def cluster_name_href(href)
+
+    def clone_completed?(args)
+      source = args[:source]
+      phase_context = args[:phase_context]
```
this is statemachine logic bleeding into the strategies.
can you try to keep the `phase_context` in `MiqProvision` ?

borod108 replied:
I do not know how to do this to be honest (without making a lot of ugly ifs) The phase_context is being modified in a block passed to with_provider_connection which has to be implemented in the appropriate class. Do you have a suggestion how to do it?

durandom replied:
I'm no expert on that `phase_context` as well. But I guess thats what gmcollough will complain about. Maybe some ugly ifs will do. It's better than having it here, where it really doesnt belong to.

borod108 replied:
@durandom We really do not see a way to do this, before this was part of the provisioning and it was always api specific, but we had only one api. Now we have two.
This part is different between the versions so to keep the same logic we need to deal with this separately.

durandom commented on app/models/manageiq/providers/redhat/infra_manager/ovirt_services/strategies/v3.rb:
```diff
@@ -19,5 +25,185 @@ def cluster_name_href(href)
+      end
+    end
+
+    class GeneralUpdateMethodNamesDecorator < SimpleDelegator
```
why not create a file for this class? then it can also be autoloaded (from specs e.g.)

borod108 replied:
Ok, should I move all classes I created to separate files? They are all used only in this specific context as decorators.

durandom replied:
nah, how you like. It just came to my mind :)

durandom commented on app/models/manageiq/providers/redhat/infra_manager/ovirt_services/strategies/v3.rb:
```diff
@@ -19,5 +25,185 @@ def cluster_name_href(href)
+      vm.with_provider_object(&:suspend)
+    end
+
+    class NicsDecorator < SimpleDelegator
```
also here. why not create a file for this class?

durandom commented on app/models/manageiq/providers/redhat/infra_manager/vm/operations.rb:
```diff
@@ -1,6 +1,7 @@
 module ManageIQ::Providers::Redhat::InfraManager::Vm::Operations
   extend ActiveSupport::Concern
 
+  require 'ovirt'
```
why here?

durandom commented on config/settings.yml:
```diff
@@ -101,7 +101,7 @@
       :read_timeout: 60
   :ems_redhat:
-    :use_ovirt_engine_sdk: false
+    :use_ovirt_engine_sdk: true
```
I thought we default to `false`?

borod108 replied:
Sorry my bad, I was pushing some small rubocop fixes and this happened.

pkliczewski commented on app/models/manageiq/providers/redhat/infra_manager/event_parser.rb:
```diff
@@ -34,9 +34,7 @@ def self.event_to_hash(event, ems_id = nil)
-      ems.with_provider_connection do |rhevm|
-        username = Ovirt::User.find_by_href(rhevm, user_href).try(:[], :user_name)
-      end
+      username = ems.ovirt_services.username_by_href(user_href)
```
Can you please say what you mean about data transformation? The event structure requires username and we get in the event something like:

`<user href="/api/users/0000001a-001a-001a-001a-0000000002b5" id="0000001a-001a-001a-001a-0000000002b5"/>`

durandom replied:
re-arranging the data in the event and preparing it for MiqEvent to store.

but that might get easier once the graph refresh is used - Ladas has some nice examples where he uses event data

durandom commented:
ok, except for the `phase_context` stuff. But if you dont find a way to refactor it, we can try to get it in like this...

pkliczewski commented on app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb:
```diff
@@ -128,7 +130,7 @@ class HostPreloadedAttributesDecorator < SimpleDelegator
         @obj = host
         @nics = connection.follow_link(host.nics)
-        @statistics = connection.follow_link(host.statistics)
+        @statistics = connection.follow_link(host.statistics) if host.statistics
```
This check won't work. Host.statistics returns an array of Statistic. The follow_link method expects to have href attribute on the object passed which array do no have. It seems that we need to iterate over host.statistics and call follow_link for each statistic there is.

jhernand replied:
That is a bug in the oVirt engine, which has been already fixed:

restapi: Don't add host statistics by default
https://gerrit.ovirt.org/74925

The workaround is to check if `host.statistics` is a link before actually using it:

```ruby
@statistics = connection.link?(host.statistics) ? connection.follow_link(host.statistics) : host.statistics
```

borod108 replied:
thank you! done.

pkliczewski commented on app/models/manageiq/providers/redhat/infra_manager/inventory/strategies/v4.rb:
```diff
@@ -98,7 +98,7 @@ def collect_templates
 
     def search_templates(search)
-      connection.system_service.templates_service.list(:search => search).collect do |template|
+      conneion.system_service.templates_service.list(:search => search).collect do |template|
```
We need to fix it. I looked for it before but thought it was fixed. Now I found it.

durandom commented:
@borod108 maybe you want to squash all commits into one? Seems there are just leftovers and and rubocop fixes...

Also - are we waiting for QE tests to pass on this PR or is it good to go from your side?

@miq-bot add_label fine/yes

borod108 commented:
@durandom so we fixed the rubocops, squashed the commits, and stabilized the version (for both apis with accent on verifying version 3).
We feel ready to merge.

durandom commented on spec/support/examples_group/shared_examples_for_provisioning_state_machine.rb:
```diff
@@ -1,6 +1,7 @@
 shared_examples_for "common rhev state machine methods" do
   it "#customize_destination" do
     allow(@task).to receive(:get_provider_destination).and_return(nil)
+    allow(@task).to receive(:destination_image_locked?).and_return(false)
```
This is the only part that will be problematic after the split.
You will need to extract this shared example to a shared example in the new repo

durandom replied:
@borod108 @pkliczewski this will cause merge conflicts. Just inline those shared_examples into the specs. This will make backporting much easier...

gmcculloug commented:
IMHO If we are close to extracting the RHV repo it would make more sense to do that an apply this PR there instead of trying to get this merged quickly.  With other activities going on and some people on PTO I do not see that this PR could be properly reviewed quickly.  One commit with 35 files only extends the timeframe required.

durandom commented:
@gmcculloug fair enough, so we'll split repos and ping you again on the new PR

durandom commented:
@borod108 this can be closed

gmcculloug commented:
PR moved to new provider repo https://github.com/ManageIQ/manageiq-providers-ovirt/pull/2
